cmd/agent/entrypoint.go

```diff
@@ -249,6 +252,74 @@ func (ep *Entrypoint) getReporterMetrics() map[string]interface{} {
 	return map[string]interface{}{"enabled-features": ep.cfg.EnabledFeatures}
 }
 
+func (ep *Entrypoint) supportHandler(rw http.ResponseWriter, r *http.Request) {
+	duration := ep.srv.HTTPServer.WriteTimeout.Seconds()
+	if r.URL.Query().Has("duration") {
+		durationInt, err := strconv.Atoi(r.URL.Query().Get("duration"))
+		if err != nil {
+			http.Error(rw, fmt.Sprintf("duration value (in seconds) should be a positive integer: %s", err), http.StatusBadRequest)
+			return
+		}
+		if durationInt < 1 {
+			http.Error(rw, "duration value (in seconds) should be larger than 1", http.StatusBadRequest)
+			return
+		}
+		duration = float64(durationInt)
+		if duration > ep.srv.HTTPServer.WriteTimeout.Seconds() {
+			http.Error(rw, "duration value exceeds the server's write timeout", http.StatusBadRequest)
+			return
+		}
+	}
+
+	// Keep track of current loggers.
+	// TODO(@tpaschalis) Can we also dynamically up the log_level while the
+	// support bundle is being generated?
+	ep.mut.Lock()
+	var (
+		srvLogger     = ep.log.GetLogger()
+		metricsLogger = ep.promMetrics.GetLogger()
+		logsLogger    = ep.lokiLogs.GetLogger()
+		tracesLogger  = ep.tempoTraces.GetLogger()
+
+		cfg             = ep.cfg
+		enabledFeatures = ep.cfg.EnabledFeatures
+		httpSrvAddress  = ep.srv.HTTPAddress().String()
+	)
+	ep.mut.Unlock()
+
+	// Create new substitute loggers that tee off to a threadsafe buffer.
+	logfmtLogger, zapLogger, logsBuffer := supportbundle.GetSubstituteLoggers(cfg.Server.LogLevel.Logrus, tracesLogger)
+
+	// Don't forget to restore the original loggers afterwards.
+	defer func() {
+		ep.log.SetLogger(srvLogger)
+		ep.promMetrics.SetLogger(metricsLogger)
+		ep.lokiLogs.SetLogger(logsLogger)
+		ep.tempoTraces.SetLogger(tracesLogger)
+	}()
+
+	// Hijack subsystems loggers.
+	ep.log.SetLogger(logfmtLogger)
+	ep.promMetrics.SetLogger(logfmtLogger)
+	ep.lokiLogs.SetLogger(logfmtLogger)
+	ep.tempoTraces.SetLogger(zapLogger)
+
+	var g singleflight.Group
+	_, err, _ := g.Do("/-/support", func() (interface{}, error) {
```

Comment:
I don't think this works; it looks like you're creating a new `singleflight.Group` each time the handler is called.

Personally, I'd just have a mutex and lock it for the duration of the request.